### PR TITLE
AWS Quarkus Lambda Client

### PR DESCRIPTION
**Pull request by tpenakov**
Quarkus implementation for AWS Lambda Client

### CODE REVIEW

File: lambdaclient/runtime/src/main/java/io/quarkus/amazon/lambda/client/runtime/LambdaConfig.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * Enable using the accelerate endpoint when accessing Lambda.
+     *
+     * <p>
+     * Accelerate endpoints allow faster transfer of objects by using Amazon CloudFront's globally
+     * distributed edge locations.
+     */
+    @WithDefault("false")
+    boolean accelerateMode();
```
Comment (scrocquesel) [inline lambdaclient/runtime/src/main/java/io/quarkus/amazon/lambda/client/runtime/LambdaConfig.java:28]: remove non necessary/used property for this service

File: devservices/lambda/src/main/java/io/quarkus/amazon/devservices/lambda/LambdaDevServicesProcessor.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    @SuppressWarnings("unused")
+    public void createFunctions(LocalStackContainer localstack, LambdaDevServiceCfg configuration) {
+        try (LambdaClient client = LambdaClient.builder()
+                .endpointOverride(localstack.getEndpointOverride(Service.LAMBDA))
+                .region(Region.of(localstack.getRegion()))
+                .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials
+                        .create(localstack.getAccessKey(), localstack.getSecretKey())))
+                .httpClientBuilder(UrlConnectionHttpClient.builder())
+                .build()) {
+            client.createFunction(b -> b.functionName(HELLO_LAMBDA)
```
Comment (scrocquesel) [inline devservices/lambda/src/main/java/io/quarkus/amazon/devservices/lambda/LambdaDevServicesProcessor.java:49]: Should be moved to either `test` or `integration-tests` (preferably the latter in this case, as the class will be much empty, similar to SSM).

File: integration-tests/src/main/java/io/quarkus/it/amazon/lambda/LambdaResource.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @GET
+    @Path("blocking")
+    @Produces(TEXT_PLAIN)
+    public String testBlocking() {
+        LOG.info("Testing Blocking Lambda client with lambda: " + LAMBDA);
+
+        try {
+            final InvokeResponse response = lambdaUtils.invokeLambda(functionInfo);
```
Comment (scrocquesel) [inline integration-tests/src/main/java/io/quarkus/it/amazon/lambda/LambdaResource.java:56]: Simplify the test, we just need to assert the client can actually talk to the aws control plane and not testing the lambda itself. I feel like the test will take too much times.

**Review by scrocquesel — CHANGES_REQUESTED** [review]
Thank you for your contribution; it is warmly welcomed. I have conducted a quick review. Do not hesitate to request a review when you feel ready.

File: lambdaclient/runtime/pom.xml [hunk 4]
```diff
@@ -0,0 +1,105 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.quarkiverse.amazonservices</groupId>
+        <artifactId>quarkus-amazon-lambda-client-parent</artifactId>
```
[the remark this answers is not in the document]
Comment (tpenakov) [reply lambdaclient/runtime/pom.xml:9]: Addressed

File: lambdaclient/runtime/src/main/java/io/quarkus/amazon/lambda/client/runtime/LambdaConfig.java (hunk 1 above)
Comment (tpenakov) [reply lambdaclient/runtime/src/main/java/io/quarkus/amazon/lambda/client/runtime/LambdaConfig.java:28]: Addressed

File: integration-tests/src/main/java/io/quarkus/it/amazon/lambda/LambdaResource.java (hunk 3 above)
Comment (tpenakov) [reply integration-tests/src/main/java/io/quarkus/it/amazon/lambda/LambdaResource.java]: Addressed

File: devservices/lambda/src/main/java/io/quarkus/amazon/devservices/lambda/LambdaDevServicesProcessor.java (hunk 2 above)
Comment (tpenakov) [reply devservices/lambda/src/main/java/io/quarkus/amazon/devservices/lambda/LambdaDevServicesProcessor.java:49]: Addressed

Comment (tpenakov) [reply devservices/lambda/src/main/java/io/quarkus/amazon/devservices/lambda/LambdaDevServicesProcessor.java]: Moved to IT

**Review by scrocquesel — CHANGES_REQUESTED** [review]
made a PR https://github.com/tpenakov/quarkus-amazon-services/pull/1

**Comment by tpenakov** [comment]
@scrocquesel - PR is approved and merged

File: lambda/deployment/src/test/java/io/quarkus/amazon/lambdaclient/deployment/LambdaDevServicesTest.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    public static final String HELLO_LAMBDA = "hello-lambda";
+    public static final String BODY = "body";
+    @Inject
+    Instance<LambdaClient> client;
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addAsResource(new StringAsset("quarkus.aws.devservices.localstack.image-name=localstack/localstack:2.3.0"),
```
Comment (scrocquesel) [inline lambda/deployment/src/test/java/io/quarkus/amazon/lambdaclient/deployment/LambdaDevServicesTest.java:29]:
```suggestion
                    .addAsResource(new StringAsset("quarkus.aws.devservices.localstack.image-name=localstack/localstack:3.0.1"),
```

**Review by scrocquesel — APPROVED** [review]

**Comment by scrocquesel** [comment]
@all-contributors please add @tpenakov  for code

**Comment by scrocquesel** [comment]
@tpenakov Thank you for your contribution

**Comment by tpenakov** [comment]
Hey @scrocquesel ,
I am glad that I finally managed to contribute to my favorite framework.
Do you want me to help you with creating an example application?
For me, it is important to have a sample where the other lambda can be called from `LambdaClient`?

**Comment by tpenakov** [comment]
@scrocquesel - there might be a maven artifact name collision with this [AWS Lambda](https://quarkus.io/guides/aws-lambda)?
What do you think?

**Comment by scrocquesel** [comment]
> @scrocquesel - there might be a maven artifact name collision with this [AWS Lambda](https://quarkus.io/guides/aws-lambda)?
> What do you think?
> 

Not the same package group per see

**Comment by scrocquesel** [comment]
> Hey @scrocquesel ,
> I am glad that I finally managed to contribute to my favorite framework.
> Do you want me to help you with creating an example application?
> For me, it is important to have a sample where the other lambda can be called from `LambdaClient`?
> 

@tpenakov you may create a quick start project https://github.com/quarkusio/quarkus-quickstarts

This is usually this quickstart we use for the  documentation with a step by step explanation.

**Comment by tpenakov** [comment]
> > @scrocquesel - there might be a maven artifact name collision with this [AWS Lambda](https://quarkus.io/guides/aws-lambda)?
> > What do you think?
> 
> Not the same package group per see

Hey @scrocquesel ,
Based on [Amazon Lambda Client documentation](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-lambda.html) when you create a project you are using the following maven command:
```
mvn io.quarkus.platform:quarkus-maven-plugin:3.6.0:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=amazon-lambda-quickstart \
    -DclassName="org.acme.lambda.QuarkusLambdaSyncResource" \
    -Dpath="/sync" \
    -Dextensions="resteasy-reactive-jackson,amazon-lambda"
cd amazon-lambda-quickstart
```
I am not certain which artifact will be used here, but it is misleading in all cases.

**Comment by tpenakov** [comment]
> > > > @scrocquesel - there might be a maven artifact name collision with this [AWS Lambda](https://quarkus.io/guides/aws-lambda)?
> > > > What do you think?
> > > 
> > > 
> > > Not the same package group per see
> > 
> > 
> > Hey @scrocquesel , Based on [Amazon Lambda Client documentation](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-lambda.html) when you create a project you are using the following maven command:
> > ```
> > mvn io.quarkus.platform:quarkus-maven-plugin:3.6.0:create \
> >     -DprojectGroupId=org.acme \
> >     -DprojectArtifactId=amazon-lambda-quickstart \
> >     -DclassName="org.acme.lambda.QuarkusLambdaSyncResource" \
> >     -Dpath="/sync" \
> >     -Dextensions="resteasy-reactive-jackson,amazon-lambda"
> > cd amazon-lambda-quickstart
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am not certain which artifact will be used here, but it is misleading in all cases.
> 
> @tpenakov You are right. It fails with multiple matches. Should be `io.quarkiverse.amazonservices:quarkus-amazon-lambda` to avoid any conflict. Would you mind create an issue or open a PR ?

Ok. 
I will create a new PR...

**Comment by scrocquesel** [comment]
@tpenakov we will have the same issue with the server extension. It will match both. Yes, this is unfortunate.
@gsmet WDYT ? Regarding amazon-lambda, we now have two extensions with the same short name. One from quarkiverse and one from quarkus. Should we trick the convention used for all other clients or modify documentation of both extension to use the fqn ?

**Pull request merged**